compiler/circle-mpqsolver/src/Methods/BisectionSolver.cpp

```diff
@@ -0,0 +1,291 @@
+  auto graph = module->graph(0);
+  auto active_nodes = loco::active_nodes(loco::output_nodes(graph));
+  std::set<loco::Node *> in_out_nodes;
+  {
+    auto nodes = loco::input_nodes(graph);
+    in_out_nodes.insert(std::begin(nodes), std::end(nodes));
+  }
+  {
+    auto nodes = loco::output_nodes(graph);
+    in_out_nodes.insert(std::begin(nodes), std::end(nodes));
+  }
+
+  bool int16_at_input =
+    error_at_input_is_larger_than_at_output(nodes_depth, 0.5f * (max_depth + min_depth));
+
+  // float min_depth_acc = int8_accuracy;
+  // float max_depth_acc = int16_accuracy;
+  while (true)
+  {
+    // float cut_depth =
+    //   ComputeLinearRoot(_accuracy, min_depth, max_depth, min_depth_acc, max_depth_acc);
+    float cut_depth = 0.5f * (min_depth + max_depth);
+
+    if (last_depth == int(cut_depth))
+    {
+      break;
+    }
+    last_depth = int(cut_depth);
+
+    auto nodes = graph->nodes();
+    LayerParams layer_params;
+    for (auto &node : active_nodes)
+    {
+      auto cur_node = loco::must_cast<luci::CircleNode *>(node);
+      if (in_out_nodes.find(cur_node) != in_out_nodes.end())
+      {
+        continue;
+      }
+      auto iter = nodes_depth.find(cur_node);
+      if (iter == nodes_depth.end())
+      {
+        continue;
+      }
+
+      float depth = iter->second;
+      if ((depth <= cut_depth && int16_at_input) || (depth >= cut_depth && !int16_at_input))
+      {
+        auto layer_param = std::make_shared<LayerParam>();
+        {
+          layer_param->name = cur_node->name();
+          layer_param->dtype = "int16";
+          layer_param->granularity = "channel";
+        }
+
+        layer_params.emplace_back(layer_param);
+      }
+    }
+
+    float cur_accuracy = evaluator.Evaluate("uint8", layer_params);
+    std::cerr << cut_depth << " : " << cur_accuracy << std::endl;
+
+    if (cur_accuracy < _accuracy)
+    {
+      int16_at_input ? (max_depth = cut_depth) : (min_depth = cut_depth);
+      //  int16_at_input ? (max_depth_acc = cur_accuracy) : (min_depth_acc = cur_accuracy);
+      best_params = layer_params;
+      best_depth = cut_depth;
+    }
+    else
+    {
+      int16_at_input ? (min_depth = cut_depth) : (max_depth = cut_depth);
+      //  int16_at_input ? (min_depth_acc = cur_accuracy) : (max_depth_acc = cur_accuracy);
+    }
+  }
+
+  if (best_params.empty())
+  {
+    std::cerr << "Failed to optimal configuration" << std::endl;
+    return EXIT_FAILURE;
+  }
+
+  std::cerr << "Found optimal configuration at " << best_depth << " depth." << std::endl;
+  _module = evaluator.Quantize("uint8", best_params);
+
+  return EXIT_SUCCESS;
+}
+
+void BisectionSolver::save(const std::string &output_model_path)
+{
+  // Export to output Circle file
+  luci::CircleExporter exporter;
+
+  luci::CircleFileExpContract contract(_module.get(), output_model_path);
+
+  if (!exporter.invoke(&contract))
+  {
+    throw std::runtime_error("Failed to export '" + output_model_path + "'");
+  }
+}
```

Comment:
This function shouldn't be here for SRP as well.

Reply:
Ok.